sql-plugin/src/main/scala/com/nvidia/spark/rapids/RmmRapidsRetryHelper.scala

```diff
@@ -0,0 +1,198 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids
+
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+
+import com.nvidia.spark.rapids.RapidsPluginImplicits._
+import com.nvidia.spark.rapids.jni.{RetryOOM, RmmSpark, SplitAndRetryOOM}
+
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+/**
+ * RmmRapidsRetryHelper is used to attempt work on the GPU with the ability to retry it,
+ * abstracting the retry logic from the calling code.
+ * @param input a SpillableColumnarBatch that we would like to attempt work with
+ * @param splitPolicy a function that can split a SpillableColumnarBatch into multiple
+ *                    spillable batches, can be set to null in case splits are not handled.
+ * @param mergePolicy a function that can be used to merge SpillableColumnarBatches that had
+ *                    been previously split, can be set to null in case merging is not handled
+ *                    or not required.
+ * @param maxRetries the maximum amount of times the retry loop will re-attempt on RetryOOM or
+ *                   SplitAndRetryOOM (3 retries by default)
+ */
+class RmmRapidsRetryHelper(
+    input: SpillableColumnarBatch,
+    splitPolicy: SpillableColumnarBatch => Seq[SpillableColumnarBatch],
+    mergePolicy: Seq[SpillableColumnarBatch] => SpillableColumnarBatch,
+    maxRetries: Int = 3) extends Arm {
+  // TODO: in the future we probably want to take this from the SpillableColumnarBatch
+  private val inputSpillCallback = RapidsBuffer.defaultSpillCallback
+  private var numRetries = 0
+  private var doSplit = false
+  private val attemptStack = new mutable.ArrayStack[SpillableColumnarBatch]()
+
+  /**
+   * withRetry takes a function `fn` that takes as input a `ColumnarBatch` and outputs
+   * a `ColumnarBatch`, in other words, the caller doesn't own the spillables in `input`.
+   * The ownership allows `withRetry` to retry without worrying about OOM exceptions
+   * in `fn`, but it means we should use `withRetry` at fine granularity, around a cuDF
+   * function call.
+   *
+   * As `fn` succeeds, the results are made spillable and queued. This, currently,
+   * requires calling `contiguousSplit` and doubling the memory usage, and so we could OOM
+   * while making the batch spillable, which is OK since we would get a RetryOOM or
+   * SplitOrRetryOOM again.
+   *
+   * @param fn function to process a batch in an OOM-safe way ColumnarBatch => ColumnarBatch
+   * @return Seq[SpillableColumnarBatch] for successful results
+   */
+  def withRetry(fn: ColumnarBatch => ColumnarBatch): Seq[SpillableColumnarBatch] = {
```

Comment:
I don't want to assume that we should keep all of the output in memory before going on with processing. I can see that this can be helpful to have minimal changes to existing code.  But, I would much rather see this as a  `WithRetryIterator with Iterator[ColumnarBatch]` that would wrap either an input Iterator of `SpillableColumnarBatch` or `ColumnarBatch` if the input is `ColumnarBatch` it would make it spillable before doing the processing. It would handle the retry and the split if it is needed.  We can worry about combining smaller batches into larger ones later. The only reason I would see to do that would be because we are trying to fit this new processing into existing code and don't want to make a lot of changes to it.

That is what I think the base API should look like and if we want to build an API like this on top of it, I am okay with it. But we would want follow on issues to fix every place that we are not using the Iterator API.

Reply:
+1 let me tinker with it to get what we want and will ask for more eyes on it.